extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/service/QuarkusMutableIdentifierGeneratorFactory.java

```diff
@@ -0,0 +1,82 @@
+package io.quarkus.hibernate.orm.runtime.service;
+
+import java.io.Serializable;
+import java.util.Properties;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.hibernate.dialect.Dialect;
+import org.hibernate.id.IdentifierGenerator;
+import org.hibernate.id.factory.internal.DefaultIdentifierGeneratorFactory;
+import org.hibernate.id.factory.spi.MutableIdentifierGeneratorFactory;
+import org.hibernate.service.spi.ServiceRegistryAwareService;
+import org.hibernate.service.spi.ServiceRegistryImplementor;
+import org.hibernate.type.Type;
+
+/**
+ * Wraps the default DefaultIdentifierGeneratorFactory so to make sure we store the Class references
+ * of any IdentifierGenerator which is accessed during the build of the Metadata.
+ *
+ * This is not to register them for reflection access: all reflective instantiation is performed
+ * during the build of the Metadata and is therefore safe even in native mode;
+ * however we still need the Class instances as some runtime operations will need these, and
+ * will look them up by either fully qualified name (and then reflection) or strategy name.
+ *
+ * Since all IdentifierGenerator types used by a model are accessed during the Metadata creation,
+ * just watching for these will provide the full list of Class instances we need to keep.
+ */
+public final class QuarkusMutableIdentifierGeneratorFactory
+        implements MutableIdentifierGeneratorFactory, Serializable, ServiceRegistryAwareService {
+
+    private final DefaultIdentifierGeneratorFactory original = new DefaultIdentifierGeneratorFactory();
+    private final ConcurrentHashMap<String, Class<? extends IdentifierGenerator>> typeCache = new ConcurrentHashMap<>();
```

Comment:
Just wondering if we should clear this cache at some point, considering we keep references to the classes? Or is it taken care of somehow?

Reply:
No we don't want to clear them as they are used by ORM at runtime:
- the instance is scoped for the specific SessionFactgory instance
- It will only contain references to Class instance of the few ID generators actually used (stuff which is going to be on the heap even if we don't refer to it)